### PR TITLE
Make it able to throw in Warden::Manager.on_request

### PR DESCRIPTION
**Pull request by godfat**
In `Warden::Test::Helpers#login_as`, it's using
`Warden.on_next_request`, which is using
`Warden::Manager.on_request`, and it would call
`proxy.set_user` which would possibly throw.

This means before this patch, `Proxy.new` could
throw in the tests. We're also expecting `env['warden']`
to be set in `Warden::Manager`, therefore we need to
decouple initializing the object and calling the `on_request`
callbacks, and move `env['warden'].on_request` inside the
`catch` block to handle the logging in process properly
in the tests.

I am not sure if people are relying on this behaviour
other than tests, but I assume it's not bad to be able to
`throw` in `Warden::Manager.on_request` as a way to
stop the chain early.

Without this patch, in the tests, `:warden` would be thrown
whenever it failed to login. People might be depending on
this behaviour, but throwing symbols like this could break
people's cleanup code.

Here's an example: https://github.com/rails/rails/pull/28373

Rails was using `rescue Exception` before to cleanup stuffs,
but it would not be rescuing any symbols. There were other
codes doing things like this, causing some weird transient
test failures like this for GitLab:
https://gitlab.com/gitlab-org/gitlab-ce/issues/40093

In short, we shouldn't let any `throw` bubble up. The other
way to fix this would be not using `Warden::Manager.on_request`
to login, but that might need more changes and could break
even more codes. Also, I think it's not bad to be able to
`throw :warden` in `Warden::Manager.on_request`.

/cc @Reprazent

### CODE REVIEW

File: lib/warden/proxy.rb [hunk 1]
```diff
@@ -29,6 +29,11 @@ def initialize(env, manager) #:nodoc:
       @env, @users, @winning_strategies, @locked = env, {}, {}, false
       @manager, @config = manager, manager.config.dup
       @strategies = Hash.new { |h,k| h[k] = {} }
+    end
+
+    # Run the on_request callbacks
+    # :api: public
+    def on_request
```
Comment (Reprazent) [inline lib/warden/proxy.rb]: I think this can be private, but I'll leave that up to the maintainers.

File: lib/warden/test/mock.rb [hunk 2]
```diff
@@ -37,7 +37,7 @@ def env
       def app
         @app ||= begin
           opts = {
-            failure_app: lambda {
+            failure_app: lambda { |env|
```
Comment (Reprazent) [inline lib/warden/test/mock.rb]: Do we need the `env` here?

Comment (godfat) [reply lib/warden/test/mock.rb]: Yes, in newer Ruby, lambda would strictly check the arity. Without this change, I'll need to provide a `failure_app` in the throwing test because it's using this default `failure_app`, which needs to take an `env`.

File: lib/warden/proxy.rb (hunk 1 above)
Comment (godfat) [reply lib/warden/proxy.rb]: At first I want to make it private, so that I could be lazy without writing tests, docs, etc. However if we're making this private, we probably need to write like `env['warden'].__send__(:on_request)`.

I am ok with this, but I am not sure if it's fine. (We don't have friends in Ruby 😓)

File: lib/warden/test/mock.rb (hunk 2 above)
Comment (godfat) [reply lib/warden/test/mock.rb]: I just realized that I should probably name this `_e` to match the app here. Let me know if I need to rename it.

File: lib/warden/proxy.rb (hunk 1 above)
Comment (Reprazent) [reply lib/warden/proxy.rb]: I meant `:api: private` in this case, and then naming the method `#_on_request`

I think that's how the methods that aren't part of the public api of warden are indicated. But I'm not sure that's the actual convention in this project 😄.

Comment (godfat) [reply lib/warden/proxy.rb]: Oh! You're right, I didn't think of that it could be `:api: private` yet it's a public method in Ruby. I'll make the change, but maybe still call it `on_request` because the other parts weren't using `_` as the prefix, and those with them have a counterpart method.

File: lib/warden/test/mock.rb (hunk 2 above)
Comment (godfat) [reply lib/warden/test/mock.rb]: I renamed this while I am at it.

**Comment by godfat** [comment]
Is there anything I can do to push this forward?
